crates/price-feed/src/price_provider/mod.rs

```diff
@@ -0,0 +1,78 @@
+use alloy_primitives::Address;
+use rust_decimal::Decimal;
+use serde::{Deserialize, Serialize};
+use time::OffsetDateTime;
+
+pub(crate) mod dia;
+pub(crate) mod erc4626;
+pub(crate) mod pyth;
+
+/// This is the struct that we expect to receive at `https://api.diadata.org/v1/assetQuotation/`.
+#[derive(Clone, Debug, Deserialize)]
+pub struct PriceInfoFromProvider {
+    #[serde(rename = "Price")]
+    pub token_price: Decimal,
+    #[serde(
+        rename = "Time",
+        deserialize_with = "time::serde::iso8601::deserialize"
+    )]
+    pub time: OffsetDateTime,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum PriceFetchError {
+    #[error("Reqwest error: {0}")]
+    Reqwest(#[from] reqwest::Error),
+
+    #[error("Unexpected pyth response")]
+    UnexpectedPythResponse,
+
+    #[error("Provider error: {0}")]
+    Provider(String),
+
+    #[error("Contract call error: {0}")]
+    ContractCall(#[from] alloy_contract::Error),
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub enum PriceProvider {
+    Dia(String),
+    Pyth(String),
+    Static(Decimal),
+    ERC4626 {
+        underlying_price_provider: Box<PriceProvider>,
```

Comment:
Is this really a recursive structure? From the definition, it might suggest we have a `LinkedList` of `PriceProvider`s. It is a neat use of Rust's algerbraic types, but I wonder whether we really need recursive types.